### PR TITLE
[RFC] Making working with LLVM a bit more convinient

### PR DESCRIPTION
**Pull request by vchuravy**
These are two changes that I have been carrying around locally for a while.

1. Allow `JULIA_LLVM_ARGS` in non-debug builds. Since debug builds and release builds
   differ sometimes, it can be quite useful to also pass them to release builds.

2. Make `LLVM_DEBUG` a three value flag:
   1. `LLVM_DEBUG=0` implies LLVM_CMAKE_BUILDTYPE=Release
   2. `LLVM_DEBUG=1` implies LLVM_CMAKE_BUILDTYPE=Debug
   3. `LLVM_DEBUG=Release` implies LLVM_CMAKE_BUILDTYPE=RelWithDebInfo
   The last option is quite helpful since it gives one most of the information,
   but is not as slow as a full debug build. (One generally wants to pair that with `LLVM_ASSERTIONS=1`)

Also `LLVM_ASSERTIONS=1` is no longer necessary for debug builds since `3.9` when using cmake.

### CODE REVIEW

File: doc/src/devdocs/llvm.md [hunk 1]
```diff
@@ -48,17 +48,22 @@ LLVM_VER = 3.5.0
...
+latter will suffice and it quite a bit faster. If you use `LLVM_DEBUG = Release` you will also
+want to set `LLVM_ASSERTIONS = 1` to enable diagonstics for different passes. Only `LLVM_DEBUG = 1`
+implies that option by default.
+
 ## Passing options to LLVM
 
-You can pass options to LLVM using *debug* builds of Julia. To create a debug build, run `make debug`.
-The resulting executable is `usr/bin/julia-debug`. You can pass LLVM options to this executable
-via the environment variable `JULIA_LLVM_ARGS`. Here are example settings using `bash` syntax:
+You can pass LLVM options to via the environment variable `JULIA_LLVM_ARGS`.
```
Comment (vtjnash) [inline doc/src/devdocs/llvm.md]: word order

**Review by vtjnash — APPROVED** [review]

**Comment by iblis17** [comment]
> Also `LLVM_ASSERTIONS=1` is no longer necessary for debug builds since `3.9` when using `cmake`.

Does `LLVM_CMAKE_BUILDTYPE=Debug` imply `LLVM_ASSERTIONS=1` ?

**Comment by vchuravy** [comment]
Yes it does nowadays: https://github.com/llvm-mirror/llvm/blob/a093ef43dd592b729da46db4ff3057fef9a46023/CMakeLists.txt#L300-L304

**Comment by iblis17** [comment]
> Yes it does nowadays

I think we should update the instruction in CI scripts as well.

**Comment by vchuravy** [comment]
> I think we should update the instruction in CI scripts as well.

Why? We don't build CI with debug info and the assertions there are generally useful. Are you suggesting that we should build with `RelWithDebInfo`?

**Comment by iblis17** [comment]
We have LLV_ASSERTION in Travis's script, Circle CI's script and BSD's
e.g. https://github.com/JuliaLang/julia/blob/master/.travis.yml#L73

Is it okay to switch to `LLVM_DEBUG=1`?

**Comment by vchuravy** [comment]
`LLVM_DEBUG=1` is **a lot** slower, it is basically unusable for anything, except if you have a very precise reproducer. `RelWithDebInfo` is faster, but since we are at the time limit of most CI systems, I wouldn't want to add any more delay, especially since we rarely use the debug information on the CI system itself.

`LLVM_ASSERTIONS=1` is great since it has minimal overhead and does the job we need it to do, e.g. check that Julia code doesn't invalidate LLVM invariants. So in my opinion we shouldn't change anything on the CI systems, and these changes are primarily meant for developers.

File: doc/src/devdocs/llvm.md (hunk 1 above)
Comment (vtjnash) [reply doc/src/devdocs/llvm.md]: drop `to`

**Review by vtjnash — APPROVED** [review]

**Review by staticfloat — APPROVED** [review]

**Pull request merged**